src/main/java/com/cloudbees/groovy/cps/impl/DGMPatcher.java

```diff
@@ -0,0 +1,314 @@
+ * <p>
+ * {@link DefaultGroovyMethods} are build-time processed (where?) to a series of "dgm" classes, and this gets
+ * loaded into {@link MetaClass} structures in {@link GeneratedMetaMethod.DgmMethodRecord#loadDgmInfo()}.
+ * <p>
+ * The code above is called from {@link MetaClassRegistryImpl#MetaClassRegistryImpl(int,boolean)} , which
+ * uses {@link CachedClass#setNewMopMethods(List)} to install method definitions from DGM.
+ * {@link CachedClass#setNewMopMethods(List)}  internally calls into {@link CachedClass#updateSetNewMopMethods(List)},
+ * which simply updates {@link ClassInfo#newMetaMethods}. This is where the method definitions stay for a while
+ * <p>
+ * The only read usage of {@link ClassInfo#newMetaMethods} is in {@link CachedClass#getNewMetaMethods()}, and
+ * this method builds its return value from all the super types. This method is then further used by
+ * {@link MetaClassImpl} when it is instantiated and build its own index.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+class DGMPatcher {
+    // we need to traverse various internal fields of the objects
+    private final Field MetaClassImpl_myNewMetaMethods = field(MetaClassImpl.class,"myNewMetaMethods");
+    private final Field MetaClassImpl_newGroovyMethodsSet = field(MetaClassImpl.class,"newGroovyMethodsSet");
+    private final Field MetaClassImpl_metaMethodIndex = field(MetaClassImpl.class,"metaMethodIndex");
+    private final Field ClassInfo_dgmMetaMethods = field(ClassInfo.class,"dgmMetaMethods");
+    private final Field ClassInfo_newMetaMethods = field(ClassInfo.class,"newMetaMethods");
+    private final Field ClassInfo_globalClassSet = field(ClassInfo.class,"globalClassSet");
+    private final Field ClassInfoSet_segments = field(AbstractConcurrentMapBase.class,"segments");
+    private final Field Segment_table = field(Segment.class,"table");
+
+    /**
+     * Used to compare two {@link MetaMethod} by their signatures.
+     */
+    static final class Key {
+        /**
+         * Receiver type.
+         */
+        final Class declaringClass;
+
+        /**
+         * Method name
+         */
+        final String name;
+
+        /**
+         * Method signature
+         */
+        final Class[] nativeParamTypes;
+
+        Key(Class declaringClass, String name, Class[] nativeParamTypes) {
+            this.declaringClass = declaringClass;
+            this.name = name;
+            this.nativeParamTypes = nativeParamTypes;
+        }
+
+        Key(MetaMethod m) {
+            this(m.getDeclaringClass().getTheClass(), m.getName(), m.getNativeParameterTypes());
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Key key = (Key) o;
+            return Objects.equal(declaringClass, key.declaringClass) &&
+                    Objects.equal(name, key.name) &&
```

Comment:
(Seems we are not running Animal Sniffer on this project. Would be a good idea.)